main.ts

```diff
@@ -22,68 +22,80 @@ const drawing_scale = 6;
+    const B1 = l1_constants[1];
+    const C1 = l1_constants[2];
 
     const l2_constants = get_line_constants(l2);
-    var A2 = l2_constants[0];
-    var B2 = l2_constants[1];
-    var C2 = l2_constants[2];
+    const A2 = l2_constants[0];
+    const B2 = l2_constants[1];
+    const C2 = l2_constants[2];
 
-    var det = A1 * B2 - A2 * B1;
+    const det = A1 * B2 - A2 * B1;
 
-    if (Math.abs(det) <= eps) return false;
+    if (Math.abs(det) <= eps) {
+        return {
+            intersection_exists: false,
+        };
+    }
 
-    var x = (B2 * C1 - B1 * C2) / det;
-    var y = (A1 * C2 - A2 * C1) / det;
+    const x = (B2 * C1 - B1 * C2) / det;
+    const y = (A1 * C2 - A2 * C1) / det;
 
-    return { "intersection_exists": on_segment(l1, x, y) && on_segment(l2, x, y), "intersection_point": new Vector2D(x, y) };
+    return {
+        intersection_exists: on_segment(l1, x, y) && on_segment(l2, x, y),
+        intersection_point: new Vector2D(x, y)
+    };
 }
 
-var on_segment = (line, x, y) => {
-    var X1 = line.startPosition.x;
-    var X2 = line.endPosition.x;
-    var Y1 = line.startPosition.y;
-    var Y2 = line.endPosition.y;
+function on_segment(line: Line, x: number, y: number) {
+    const X1 = line.startPosition.x;
+    const X2 = line.endPosition.x;
+    const Y1 = line.startPosition.y;
+    const Y2 = line.endPosition.y;
 
     return Math.min(X1, X2) <= x && x <= Math.max(X1, X2) &&
         Math.min(Y1, Y2) <= y && y <= Math.max(Y1, Y2);
 }
 
-var get_line_constants = (line) => {
-    var X1 = line.startPosition.x;
-    var X2 = line.endPosition.x;
-    var Y1 = line.startPosition.y;
-    var Y2 = line.endPosition.y;
+function get_line_constants(line: Line) {
+    const X1 = line.startPosition.x;
+    const X2 = line.endPosition.x;
+    const Y1 = line.startPosition.y;
+    const Y2 = line.endPosition.y;
 
-    var A = Y2 - Y1;
-    var B = X1 - X2;
-    var C = A * X1 + B * Y1;
+    const A = Y2 - Y1;
+    const B = X1 - X2;
+    const C = A * X1 + B * Y1;
     return [A, B, C];
 }
 
-const key_pressed = {
-    "left": 0,
-    "right": 0,
-    "up": 0,
-    "down": 0
-};
+const key_pressed = new Map([
+    ['left', 0],
+    ['right', 0],
+    ['up', 0],
+    ['down', 0],
+]);
 
-$(document).ready(function () {
-    const input = $('body');
-    Rx.Observable.fromEvent(input, 'keydown')
-        .map((e: KeyboardEvent) => keyDefs[e.keyCode])
-        .subscribe(e => key_pressed[e] = 1);
+$(document).ready(() => {
+    const body = $('body');
+    Rx.Observable.fromEvent(body, 'keydown')
+        .map((e: KeyboardEvent) => keyDefs.get(e.keyCode))
+        .subscribe(e => key_pressed.set(e, 1));
 
-    Rx.Observable.fromEvent(input, 'keyup')
-        .map((e: KeyboardEvent) => keyDefs[e.keyCode])
-        .subscribe(e => key_pressed[e] = 0);
+    Rx.Observable.fromEvent(body, 'keyup')
+        .map((e: KeyboardEvent) => keyDefs.get(e.keyCode))
+        .subscribe(e => key_pressed.set(e, 0));
```

Comment:
How does that work? Doesn't that return a new immutable Map?

Reply:
Nope, it's in-place. See [the docs here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/set).